File: hpcperfstats/site/frontend/src/utils/gpuClockThrottleReasons.ts
/**
 * Decode DCGM GPU clock throttle / clock event reason bitmasks
 * (dcgm_fields.h DCGM_CLOCKS_THROTTLE_REASON_* / field 112).
 *
 * Persisted metric max_gpu_clock_event_reasons stays numeric; this is display-only.
 * Blank-family and no-known-bit garbage masks return "" (never "unknown (0x…)").
 */

/**
 * DCGM FP64 blank base (dcgm_structs.h / 2**47). Also catches INT64 blank-family
 * values: INT64 blank (~2**63) is larger, but is not a JS-safe integer literal
 * (eslint no-loss-of-precision), so SPA blank checks use this threshold only.
 */
const DCGM_FP64_BLANK = 140737488355328;

function isDcgmNumericBlank(mask: number): boolean {
  if (!Number.isFinite(mask)) return false;
  return mask >= DCGM_FP64_BLANK;
}

export const DCGM_CLOCK_THROTTLE_REASON_FLAGS: ReadonlyArray<{
  bit: number;
  label: string;
}> = [
  { bit: 0x1, label: "GPU idle" },
  { bit: 0x2, label: "Application clocks setting" },
  { bit: 0x4, label: "SW power cap" },
  { bit: 0x8, label: "HW slowdown (temp/power/Pstate)" },
  { bit: 0x10, label: "Sync boost" },
  { bit: 0x20, label: "SW thermal slowdown" },
  { bit: 0x40, label: "HW thermal slowdown" },
  { bit: 0x80, label: "HW power brake" },
  { bit: 0x100, label: "Display clocks setting" },
];

/**
 * Format a DCGM clock-throttle reason bitmask as comma-separated flag names.
 * Returns "" for empty / non-finite / zero / blank / no-known-bit masks.
 */
export function formatGpuClockThrottleReasons(mask: number): string {
  if (!Number.isFinite(mask)) return "";
  if (isDcgmNumericBlank(mask)) return "";
  const intMask = Math.trunc(mask);
  if (intMask === 0) return "";

  const unsigned = intMask >>> 0;
  const parts: string[] = [];
  for (const { bit, label } of DCGM_CLOCK_THROTTLE_REASON_FLAGS) {
    if (unsigned & bit) parts.push(label);
  }
  // Fail closed: never emit unknown (0x…) for residual garbage bits.
  return parts.join(", ");
}

File: hpcperfstats/analysis/metrics/lib/gpu_clock_throttle_reasons.py
"""
Decode DCGM GPU clock throttle / clock event reason bitmasks.

Mirrors ``dcgm_fields.h`` ``DCGM_CLOCKS_THROTTLE_REASON_*`` and the SPA helper
``gpuClockThrottleReasons.ts``. Display-only; persisted metrics stay numeric.

Blank-family and no-known-bit garbage masks return ``""`` (never ``unknown
(0x…)``).

Attributes:
  DCGM_CLOCK_THROTTLE_REASON_FLAGS: ``DCGM_CLOCK_THROTTLE_REASON_FLAGS``.
  _KNOWN_BITS_MASK: ``_KNOWN_BITS_MASK``.
"""

from __future__ import annotations

from hpcperfstats.lib.dcgm_blank import is_dcgm_numeric_blank

DCGM_CLOCK_THROTTLE_REASON_FLAGS: tuple[tuple[int, str], ...] = (
    (0x1, "GPU idle"),
    (0x2, "Application clocks setting"),
    (0x4, "SW power cap"),
    (0x8, "HW slowdown (temp/power/Pstate)"),
    (0x10, "Sync boost"),
    (0x20, "SW thermal slowdown"),
    (0x40, "HW thermal slowdown"),
    (0x80, "HW power brake"),
    (0x100, "Display clocks setting"),
)

_KNOWN_BITS_MASK = 0
for _bit, _ in DCGM_CLOCK_THROTTLE_REASON_FLAGS:
  _KNOWN_BITS_MASK |= _bit


def format_gpu_clock_throttle_reasons(mask: float | int | None) -> str:
  """
  Format a DCGM clock-throttle bitmask as comma-separated flag names.
  
  Returns ``""`` for missing / non-finite / zero / blank / no-known-bit masks.
  Known bits only — residual garbage hex is never shown.
  
  Args:
    mask (float | int | None): One of ``float``, ``int``, ``None``.
  
  Returns:
    str: str produced by this call.
  
  Examples:
    >>> format_gpu_clock_throttle_reasons(None)  # doctest: +SKIP
  """
  if mask is None:
    return ""
  if is_dcgm_numeric_blank(mask):
    return ""
  try:
    value = float(mask)
  except (TypeError, ValueError):
    return ""
  if value != value:  # NaN
    return ""
  int_mask = int(value)
  if int_mask == 0:
    return ""

  # Match JS >>> 0 for non-negative display of the low 32 bits.
  unsigned = int_mask & 0xFFFFFFFF
  parts: list[str] = []
  for bit, label in DCGM_CLOCK_THROTTLE_REASON_FLAGS:
    if unsigned & bit:
      parts.append(label)
  # Fail closed: unknown residual alone or with known bits is omitted.
  return ", ".join(parts)
